GVFS/GVFS.UnitTests/Common/Database/PlaceholdersTableTests.cs

```diff
@@ -0,0 +1,435 @@
+﻿using GVFS.Common.Database;
+using GVFS.Tests.Should;
+using Moq;
+using NUnit.Framework;
+using System;
+using System.Collections.Generic;
+using System.Data;
+using System.Linq;
+
+namespace GVFS.UnitTests.Common.Database
+{
+    [TestFixture]
+    public class PlaceholdersTableTests
+    {
+        private const string DefaultPath = "test";
+        private const byte PathTypeFile = 0;
+        private const byte PathTypePartialFolder = 1;
+        private const byte PathTypeExpandedFolder = 2;
+        private const byte PathTypePossibleTombstoneFolder = 3;
+        private const string DefaultSha = "SHA1";
+
+        [TestCase]
+        public void ConstructorTest()
+        {
+            Mock<IGVFSConnectionPool> mockConnectionPool = new Mock<IGVFSConnectionPool>(MockBehavior.Strict);
+            PlaceholdersTable placeholders = new PlaceholdersTable(mockConnectionPool.Object);
+            mockConnectionPool.VerifyAll();
+        }
+
+        [TestCase]
+        public void CreateTableTest()
+        {
+            Mock<IDbCommand> mockCommand = new Mock<IDbCommand>(MockBehavior.Strict);
+            mockCommand.SetupSet(x => x.CommandText = "CREATE TABLE IF NOT EXISTS [Placeholders] (path TEXT PRIMARY KEY, pathType TINYINT NOT NULL, sha char(40) ) WITHOUT ROWID;");
+            mockCommand.Setup(x => x.ExecuteNonQuery()).Returns(1);
+            PlaceholdersTable.CreateTable(mockCommand.Object);
+            mockCommand.VerifyAll();
+        }
+
+        [TestCase]
+        public void CountTest()
+        {
+            this.TestPlaceholders(
+                (placeholders, mockCommand) =>
+                {
+                    mockCommand.SetupSet(x => x.CommandText = "SELECT count(path) FROM Placeholders;");
+                    mockCommand.Setup(x => x.ExecuteScalar()).Returns(0);
+                    placeholders.Count().ShouldEqual(0);
+                });
+        }
+
+        [TestCase]
+        public void GetAllFilePathsWithNoResults()
+        {
+            this.TestPlaceholdersWithReader(
+               (placeholders, mockCommand, mockReader) =>
+               {
+                   mockReader.Setup(x => x.Read()).Returns(false);
+                   mockCommand.SetupSet(x => x.CommandText = "SELECT path FROM Placeholders WHERE pathType = 0;");

```

Comment:
I like that the unit tests here use the literal "0". If you use the enum value in the implementation of `GetAllFilePaths()`, this test would detect a change in the specific enum values.

(We shouldn't change the enum values, but this gives us protection against it.)